### PR TITLE
[attestation] Add endorsement baseline support for oe_verify_evidence

### PR DESCRIPTION
**Pull request by anche-is-andy**
This includes the changes proposed in https://github.com/openenclave/openenclave/pull/4367

### CODE REVIEW

File: common/sgx/collateral.c
Comment (dthaler) [inline common/sgx/collateral.c:50]: This is ok for now but this should probably use an enum type in the future instead of uint8_t.

File: common/sgx/collateral.h
Comment (dthaler) [inline common/sgx/collateral.h:81]:
```suggestion
 * Caller is responsible for freeing the quote verification collateral resources
```

File: common/sgx/collateral.h
Comment (dthaler) [inline common/sgx/collateral.h:104]: 91-95 do not mention any need to free resources, nor does line 71 point to this API.

File: common/sgx/collateral.h
Comment (dthaler) [inline common/sgx/collateral.h:49]: Isn't this file a purely internal header, not published in the SDK?   https://github.com/openenclave/openenclave/blob/master/docs/ApiGuidelines.md explains:

> Experimental and internal APIs can have breaking changes across releases

Thus you should not need to deprecate these, you can just delete them.

File: common/sgx/endorsements.c
Comment (dthaler) [inline common/sgx/endorsements.c:434]: There's no reason to keep oe_create_sgx_endorsements unchanged since it's a static function in this file, you can just change it as needed, so this phrase does not make sense.

File: host/sgx/sgxquoteprovider.c
Comment (dthaler) [inline host/sgx/sgxquoteprovider.c:359]: Delete line 359, free() is defined as being safe to call with null, and it reduces code complexity and potentially increases test coverage to delete the line.

File: include/openenclave/bits/evidence.h
Comment (dthaler) [inline include/openenclave/bits/evidence.h:176]:
```suggestion
     * will be used when multiple policies of this type are given.
```
However, I think this is bug-prone, since it just encourages bad code to go undetected.
Instead, if you only use the first policy of this type, then to reduce the risk of bugs in the caller, this call should fail if the caller tries to pass multiple of them rather than silently ignoring them.

File: tests/attestation_plugin/host/host.c
Comment (dthaler) [inline tests/attestation_plugin/host/host.c:54]:
```suggestion
        // set up mocks for tests
```
fix grammar ("set up" is a verb.  "setup" is a noun)

File: tests/attestation_plugin/plugin/tests.c
Comment (dthaler) [inline tests/attestation_plugin/plugin/tests.c:789]:
```suggestion
void set_up_mocks_for_host()
```

File: tests/attestation_plugin/plugin/tests.h
Comment (dthaler) [inline tests/attestation_plugin/plugin/tests.h:35]:
```suggestion
void set_up_mocks_for_host();
```

File: common/sgx/collateral.h
Comment (dthaler) [inline common/sgx/collateral.h:87]: FYI, "params" in the name violates [OE Coding Conventions](https://github.com/openenclave/openenclave/blob/master/docs/Community/DevelopmentGuide.md) which says:

> DON'T use abbreviations unless they are already well-known terms known by users (e.g., "app", "info"), or are already required for use by developers (e.g., "min", "max", "args"). Examples of bad use would be num_widgets instead of widget_count, and opt_widgets instead of option_widgets or optional_widgets.

Unlike "args", "params" isn't something that's defined by the C language/headers and so does not fall within the exception clause.   I do see a bunch of other violations in public OE headers though so at least OE is somewhat consistent in not following its own coding conventions :(

File: enclave/sgx/collateralinfo.c
Comment (dthaler) [inline enclave/sgx/collateralinfo.c:239]: Not new in this PR so ok to leave if you want, but "buf" violates [OE coding style](https://github.com/openenclave/openenclave/blob/master/docs/Community/DevelopmentGuide.md).

File: enclave/sgx/collateralinfo.c
Comment (dthaler) [inline enclave/sgx/collateralinfo.c:249]: Not new in this PR so ok to leave if you want, but "src" violates [OE coding style](https://github.com/openenclave/openenclave/blob/master/docs/Community/DevelopmentGuide.md).

File: enclave/sgx/collateralinfo.c
Comment (dthaler) [inline enclave/sgx/collateralinfo.c:610]: Revert addition of if tests.  They increase code complexity, potentially reduce code coverage of tests, and provide no functional value here.

**Review by dthaler — COMMENTED** [review]
I am the OE API maintainer and as such, I did a review for conformance to [OE API Guidelines](https://github.com/openenclave/openenclave/blob/master/docs/ApiGuidelines.md) and [OE Coding Conventions](https://github.com/openenclave/openenclave/blob/master/docs/Community/DevelopmentGuide.md).

File: common/sgx/collateral.c
Comment (anche-is-andy) [reply common/sgx/collateral.c:50]: okay, currently there is no enum define for collateral_provider yet, need to do a refactor after this.

File: common/sgx/collateral.h
Comment (anche-is-andy) [reply common/sgx/collateral.h:49]: Cool, just realized this file is not public, is there any doc about how can identify whether a header file is public or not?

File: common/sgx/collateral.h
Comment (anche-is-andy) [reply common/sgx/collateral.h:87]: This is a very good catch, thanks! it reminds me that my mindset was sorted of limited to the name that QPL provides to OE SDK, we actually call this as baseline, now I have a better name for all of these, params sounds odd to me now

File: common/sgx/collateral.h
Comment (dthaler) [reply common/sgx/collateral.h:49]: The usual rule is:
* headers not under openenclave/include are internal
* headers under openenclave/include/openenclave/internal are also internal
* other headers under openenclave/include are public

but openenclave/include/openenclave/bits/ has a mix it seems, which is issue #4132

And of course issue #4370 is also relevant if there are internal APIs in any public headers.

Comment (anche-is-andy) [reply common/sgx/collateral.h]: Thanks, so in this sense, all edl files are also treated as public APIs

File: enclave/sgx/collateralinfo.c
Comment (anche-is-andy) [reply enclave/sgx/collateralinfo.c]: I think it's not allocated in enclave, remove it in last iteration

File: common/sgx/collateral.h
Comment (dthaler) [reply common/sgx/collateral.h]: Filed #4388

File: common/sgx/collateral.h
Comment (anche-is-andy) [inline common/sgx/collateral.h:73]: need to remove this

File: enclave/sgx/collateralinfo.c
Comment (anche-is-andy) [inline enclave/sgx/collateralinfo.c:118]: with_baseline

File: include/openenclave/internal/report.h
Comment (anche-is-andy) [inline include/openenclave/internal/report.h:18]: need to remove _t suffix

File: tests/attestation_plugin/plugin/tests.c
Comment (anche-is-andy) [inline tests/attestation_plugin/plugin/tests.c:740]: with_baseline

File: enclave/sgx/collateralinfo.c
Comment (anche-is-andy) [reply enclave/sgx/collateralinfo.c]: I would leave it as it was, there are lots of places need to be updated, better to be in a separated PR

File: common/attest_plugin.c
Comment (yentsanglee) [inline common/attest_plugin.c:318]: Use OE_RAISE_MSG() for better logging.

**Review by yentsanglee — APPROVED** [review]

**Comment by yentsanglee** [comment]
bors try

**Comment by yentsanglee** [comment]
bors try

**Comment by anche-is-andy** [comment]
bors try

**Comment by anche-is-andy** [comment]
@mingweishih  @dthaler  can you please help to review this PR again? there is a dependency for this PR

File: common/sgx/endorsements.c
Comment (mingweishih) [inline common/sgx/endorsements.c:368]: Need to validate `policies` and `policies_size`

File: host/sgx/sgxquoteprovider.c
Comment (mingweishih) [inline host/sgx/sgxquoteprovider.c:71]: Can use `OE_UINT_MAX`: https://github.com/openenclave/openenclave/blob/b08b96f25925232e4ff639389a26c71646733bc0/include/openenclave/bits/types.h#L98

**Review by mingweishih — APPROVED** [review]
Looks good. Can you squash the commits before merging?

**Comment by anche-is-andy** [comment]
> Looks good. Can you squash the commits before merging?

Sure, will do squash

**Comment by anche-is-andy** [comment]
bors try

**Comment by anche-is-andy** [comment]
bors try

**Comment by anche-is-andy** [comment]
bors retry

**Pull request merged**
